File: dotnet/csharp/src/Repo.cs
using Projector.Domain;
using System;
using System.Collections.Generic;

namespace Projector
{
    public class Repo
    {
        private readonly Dictionary<string, InventoryItem> persistence;

        public Repo() => persistence = new Dictionary<string, InventoryItem>();

        public InventoryItem Get(Guid id) => persistence.GetValueOrDefault(id.ToString()) ?? InventoryItem.Init(id);

        public void Upsert(InventoryItem inventoryItem) => persistence[inventoryItem.Id.ToString()] = inventoryItem;

        public IEnumerable<InventoryItem> GetAll() => persistence.Values;
    }
}

File: dotnet/fsharp/src/Repo.fs
module Repo

open System
open Domain

let mutable private persistence: Map<string, InventoryItem> = Map.empty;

let Get (id: Guid) =
    persistence |> Map.tryFind (id.ToString())
    |> Option.defaultValue (id |> InventoryItem.Initialize)

let Upsert (inventoryItem: InventoryItem) =
    persistence <- persistence |> Map.add (inventoryItem.Id.ToString()) inventoryItem

let GetAll () =
    persistence |> Map.toSeq |> Seq.map snd